### PR TITLE
ENH: Add time range to different plots

### LINKED ISSUE #544: Make all plots with time end at `tstop`
Currently some of the plots like `cell_response.plot_spikes_raster()` do not end at `tstop`. This change will be useful for plotting functionality in the dashboard gui.

### PR DESCRIPTION
**Pull request by aritrasinha108**
Closes #544 

This PR adds parameters for start and end simulation times limit for different plots.

### CODE REVIEW

**Comment by jasmainak** [comment]
CircleCI Is failing ...

**Comment by aritrasinha108** [comment]
> CircleCI Is failing ...

@jasmainak fixed the error. Please have a look.

File: hnn_core/viz.py [hunk 1]
```diff
@@ -1150,8 +1166,15 @@ def plot_laminar_csd(times, data, contact_labels, ax=None, colorbar=True,
     import matplotlib.pyplot as plt
     if ax is None:
         _, ax = plt.subplots(1, 1, constrained_layout=True)
-
-    im = ax.pcolormesh(times, contact_labels, np.array(data),
+    plot_data = []
+    plot_times = []
+    for time_data in data:
+        print("len(times)", len(times))
+        print("len(time_data)", len(time_data))
```
Comment (jasmainak) [inline hnn_core/viz.py:1173]: debug artifacts?

Comment (aritrasinha108) [reply hnn_core/viz.py]: Sorry, forgot to removed them while pushing. I've done it now

**Comment by aritrasinha108** [comment]
> Can you share a before and after screenshot to show how it addresses the original issue?
@jasmainak should I add a screenshot of the code that is changed and explain how the issue is addressed here or should I put the results I get after running a script that deals with it?

**Comment by jasmainak** [comment]
I mean can you share a screenshot of the plotting function you changed ... before your changes were implemented and after?

**Comment by aritrasinha108** [comment]
> I mean can you share a screenshot of the plotting function you changed ... before your changes were implemented and after?

Yeah Sure, on it now

**Comment by aritrasinha108** [comment]
## The changes that I made are listed and justified below:

## 1. `hnn_core.viz.plot_spikes_hist`

### Before:
spike hist b1

#### Inside the function, once the data and times for the trials were extracted, the function would directly move on to listing all the unique spike types and creating a mask to create labels using them.  
spike hist b2

### After:

spike hist a1

#### Before we move on to create spike labels and further process the data, we here filter or slice the `spikes_times` and `spike_types_data` using the function `_get_plot_data_trange`.  Using this method is only possible when the `data` and the `times` array are mapped one-to-one and have the same dimension. Once we receive back the `data` and `times` array, we update the `spike_times` and `spike_types_data` with the values we receive and move ahead with the function as we did earlier. 
spike hist a2


## 2. `hnn_core.viz.plot_spikes_raster`

### Before:
spike raster b1

#### Since this is an even plot, there is no data that needs to be plotted here. After grouping the cell according to their cell types, we take each cell of that type and extract their `spike_times` and add it to the `cell_type_times`, which stores the spike times for that cell type.
spike raster b2

### After:
spike raster a1

#### After extracting the spike times for each cell of a specific type, we pass the time to the `_get_plot_data_trange` function to be sliced and limited into the simulation time range. The sliced spike times array is then appended to the `cell_type_times` for subsequent operations.
spike raster a2


## 3. `hnn_core.viz.plot_laminar_csd`
### Before:
laminar csd b1

#### Since this is a color mesh plot, the  `data` here is a List of List where there is CSD data for all the channels, and each of those items here is mapped one-to-one with the `times` array.  
laminar csd b2

### After:
laminar csd a1

####  We just iterate over the CSD data of each channel here and slice out all the values which are outside the `(tmin, tmax)` range and store it inside a new List called the `plot_data` to store the filtered data to be plotted. We are also using the variable `plot_times` as `times` getting sliced in any iteration would result in the slicing operation of the next iteration becoming invalid due to differences in dimensions. 
laminar csd a2

**Comment by jasmainak** [comment]
Hi -- I am sorry if I was not clear. I don't want to see code. It is there in the diff and easier to follow there :) What I wanted to see was screenshots of the plots themselves. Also, please use markdown formatting if you really need to show code.

**Comment by aritrasinha108** [comment]
> Hi -- I am sorry if I was not clear. I don't want to see code. It is there in the diff and easier to follow there :) What I wanted to see was screenshots of the plots themselves. Also, please use markdown formatting if you really need to show code.

Oh really sorry I misunderstood. I have updated the description now. Please let me know if it's fine. thanks

**Comment by jasmainak** [comment]
@ntolley @chenghuzi do you want to weigh in here? I'm not sure if we ever want to plot tmax < tstop ... I think we only need to set the x limits using tstop. Not sure if users should be given `tmin` and `tmax` parameters.

**Comment by aritrasinha108** [comment]
> @ntolley @chenghuzi do you want to weigh in here? I'm not sure if we ever want to plot something with tmax < tstop ... I think we only need to set the x limits using tstop. So that all the plots can be aligned in subplots for the GUI.
> 
> Not sure if users should be given the option of setting `tmin` and `tmax` parameters. I find this plot somewhat misleading

@jasmainak all the other plots except the 3 on which I worked (`hnn_core.viz.plot_dipole`, `hnn_core.viz.plot_psd`,  `hnn_core.viz.plot_tfr_morlet` and `hnn_core.viz.plot_laminar_lfp`) provide the users with the option of setting `tmin` and `tmax`. 

### For eg:

Inside `hnn_core/dipole.py`, the plotting function for the `Dipole` class is defined as
```python
def plot(self, tmin=None, tmax=None, layer='agg', decim=None, ax=None,
             color='k', show=True):
       .......
       return plot_dipole(self, tmin=tmin, tmax=tmax, ax=ax, layer=layer,
                           decim=decim, color=color, show=show)
```
This function has been used in a lot of examples like `plot_connectivity.py` and others.

The same goes for other plots too.

**Comment by rythorpe** [comment]
Now that we're discussing this, I actually agree with @jasmainak that we don't really need to have `tmin` and `tmax` in any of our functions because the user can always just set the x-limits using the Matplotlib API. The only exceptions I can think of is maybe for [`plot_tfr_morlet`](https://github.com/jonescompneurolab/hnn-core/blob/b9a09d8e8f05e6462f76c119814baa716ddd23d1/hnn_core/viz.py#L575) and [`plot_psd`](https://github.com/jonescompneurolab/hnn-core/blob/b9a09d8e8f05e6462f76c119814baa716ddd23d1/hnn_core/viz.py#L716) because setting the time limits can improve compute time when mapping to the frequency domain. I think there are two paths forward that we should vote on:

1. add `tmin` and `tmax` arguements to all plot functions with a time axis in order to maintain API consistency
2. remove `tmin` and `tmax` from all plot functions *except* for `plot_tfr_morlet` and `plot_psd` and silently set the the upper limit of the time axis to `tstop`.

Are there any other options I'm missing? (@chenghuzi this may effect how the GUI renders plots so your preference probably matters the most here.)

**Comment by jasmainak** [comment]
> setting the time limits can improve compute time when mapping to the frequency domain

you may also want to leave out certain parts of the signal for the mapping, for example if there is a transient and steady-state component. You may only want to compute the psd for the steady state component. So, I agree that `tmin` and `tmax` are justified in these functions. 

I vote for 2!

**Comment by rythorpe** [comment]
@aritrasinha108 wdyt? If you have no objections, let's move forward with Option 2 where you'll still be setting the upper x-limit of the plots but with minimal user-facing complexity.

**Comment by aritrasinha108** [comment]
> @aritrasinha108 wdyt? If you have no objections, let's move forward with Option 2 where you'll still be setting the upper x-limit of the plots but with minimal user-facing complexity.

@rythorpe yeah yeah that completely works for me. Will start working on it right away

**Comment by rythorpe** [comment]
@aritrasinha108 try using `CellResponse.times` to access a list of all time samples in the simulation, then select the last element for the final time.

**Comment by aritrasinha108** [comment]
> @aritrasinha108 try using `CellResponse.times` to access a list of all time samples in the simulation, then select the last element for the final time.

oh, okay I'll do that. Thanks a lot.

**Comment by aritrasinha108** [comment]
@rythorpe @jasmainak I have made the changes you asked. Could you please check. Thanks

File: hnn_core/extracellular.py [hunk 2]
```diff
@@ -435,10 +435,6 @@ def plot_lfp(self, *, trial_no=None, contact_no=None, tmin=None, tmax=None,
             Trial number(s) to plot
         contact_no : int | list of int | slice
             Electrode contact number(s) to plot
-        tmin : float | None
-            Start time of plot in milliseconds. If None, plot entire
-            simulation.
-        tmax : float | None
             End time of plot in milliseconds. If None, plot entire simulation.
```
Comment (rythorpe) [inline hnn_core/extracellular.py:438]: Missed this line as well.

Comment (aritrasinha108) [reply hnn_core/extracellular.py]: Oh, really sorry I missed that. Removed it now

File: hnn_core/viz.py [hunk 3]
```diff
@@ -168,11 +164,11 @@ def plot_laminar_lfp(times, data, contact_labels, tmin=None, tmax=None,
         trace_offsets = np.arange(n_offsets)[:, np.newaxis] * voltage_offset
 
     for contact_no, trace in enumerate(np.atleast_2d(data)):
-        plot_data, plot_times = _get_plot_data_trange(times, trace, tmin, tmax)
-
+        plot_data = []
+        plot_times = []
```
[the remark this answers is not in the document]
Comment (aritrasinha108) [reply hnn_core/viz.py]: Sorry, my bad. Fixed it now. Oh, I'll keep that in mind from the next time. Thanks

File: hnn_core/viz.py [hunk 4]
```diff
@@ -479,11 +469,12 @@ def plot_spikes_raster(cell_response, trial_idx=None, ax=None, show=True):
             np.array(cell_response._spike_types, dtype=object)[trial_idx])
         spike_gids = np.concatenate(
             np.array(cell_response._spike_gids, dtype=object)[trial_idx])
+        tstop = spike_times[-1]
     else:
         spike_times = np.array([])
         spike_types = np.array([])
         spike_gids = np.array([])
-
+        tstop = 0
```
Comment (rythorpe) [inline hnn_core/viz.py:477]: Why is `tstop` set to zero?

**Comment by aritrasinha108** [comment]
Also, @rythorpe @jasmainak I have started with my proposal for the GSoC Project "Generalize Parameter Optimization Routines". It would be great if you could share your email IDs so that I can share the draft to the google doc for reviews and suggestions. Thanks

**Comment by ntolley** [comment]
Sorry for the late response! You can email me as well at nicholas_tolley at brown dot edu

**Comment by rythorpe** [comment]
@aritrasinha108 how's the PR going? I also wanted to give you a reminder that the deadline for GSoC proposals is fast approaching, so please reach out with a draft of your proposal as soon as possible if you're still planning on moving forward with your application. We're happy to provide feedback on your timeline, goals, etc.

File: hnn_core/viz.py [hunk 5]
```diff
@@ -479,11 +471,12 @@ def plot_spikes_raster(cell_response, trial_idx=None, ax=None, show=True):
             np.array(cell_response._spike_types, dtype=object)[trial_idx])
         spike_gids = np.concatenate(
             np.array(cell_response._spike_gids, dtype=object)[trial_idx])
+        tstop = spike_times[-1]
```
[the remark this answers is not in the document]
Comment (ntolley) [reply hnn_core/viz.py:474]: @tianqi-cheng this should use `cell_response.times[-1]`

**Comment by ntolley** [comment]
Closing in favor of #683

**Pull request closed without merging**